src/backend/commands/copy.c

```diff
@@ -722,34 +723,67 @@ CopyGetData(CopyState cstate, void *databuf, int datasize)
 	switch (cstate->copy_dest)
 	{
 		case COPY_FILE:
-			bytesread = fread(databuf, 1, datasize, cstate->copy_file);
-			if (feof(cstate->copy_file))
-				cstate->fe_eof = true;
-			if (ferror(cstate->copy_file))
 			{
-				if (cstate->is_program)
+				pgsocket 	sock = fileno(cstate->copy_file);
+				int			rc;
+
+				/*
+				 * Wait until we get data, even if interrupted by a signal,
+				 * while also checking for interrupts
+				 */
+				do
 				{
-					int olderrno = errno;
+					CHECK_FOR_INTERRUPTS();
 
-					close_program_pipes(cstate, true);
+					/* Wait until data arrives or 500 ms passes */
+#define COPY_POLL_TIMEOUT 500
+					PG_TRY();
+					{
+						rc = WaitLatchOrSocket(NULL, WL_SOCKET_READABLE | WL_TIMEOUT, 
+											   sock, COPY_POLL_TIMEOUT);
+					}
+					PG_CATCH();
+					{
+						if (cstate->is_program)
+							close_program_pipes(cstate, true);
```

Comment:
I think it is not need due to close_program_pipes_on_reset

Reply:
Oh, it was added after the original code. Okay then, removing it